### PR TITLE
GH-4991: Support for custom React toolbar widgets.

### LINKED ISSUE #4991: As an extension developer I want to add any custom React widget to the tab-bar toolbar
Regarding the UI, I had something like this in mind.
Screen Shot 2019-04-24 at 17 12 57

### PR DESCRIPTION
**Pull request by kittaakos**
Closes #4991.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

screencast 2019-04-24 17-24-56

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
@kittaakos is there an easy way I can test the functionality?

**Comment by kittaakos** [comment]
> @kittaakos is there an easy way I can test the functionality?

That's my problem with such PRs; no, unfortunately, you have to modify the code:
```ts
registry.registerItem({
    id: 'demo',
    node: <select
        name={'two'}
        value={'two'}
        title={'tooltip'}>
        {['one', 'two', 'three'].map(option => <option value={option} key={option}>{option}</option>))}
    </select>
});
```

**Comment by vince-fugnitto** [comment]
> That's my problem with such PRs; no, unfortunately, you have to modify the code:

That's fine , I just needed an example so I can test it for you :)

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx [hunk 1]
```diff
@@ -65,7 +65,7 @@ export class TabBarToolbar extends ReactWidget {
 
     protected render(): React.ReactNode {
         return <React.Fragment>
-            {[...this.items.values()].map(item => this.renderItem(item))}
+            {[...this.items.values()].map(item => TabBarToolbarItem.is(item) ? this.renderItem(item) : item.node)}
```
Comment (marechal-p) [inline packages/core/src/browser/shell/tab-bar-toolbar.tsx]: Not directly in the scope of this PR, but what is the difference between returning a `<React.Fragment>` and just returning the mapped list directly? I tried and did not notice a difference.

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx [hunk 2]
```diff
@@ -272,11 +295,13 @@ export class TabBarToolbarRegistry implements FrontendApplicationContribution {
...
      */
-    visibleItems(widget: Widget): TabBarToolbarItem[] {
+    visibleItems(widget: Widget): Array<TabBarToolbarItem | ReactTabBarToolbarItem> {
         const result = [];
         for (const item of this.items.values()) {
-            if (this.commandRegistry.isVisible(item.command, widget) &&
-                (!item.when || this.contextKeyService.match(item.when, widget.node))) {
+            const visible = TabBarToolbarItem.is(item)
+                ? this.commandRegistry.isVisible(item.command, widget)
+                : (!item.isVisible || item.isVisible(widget));
```
Comment (marechal-p) [inline packages/core/src/browser/shell/tab-bar-toolbar.tsx]: `: item.isVisible && item.isVisible(widget);` ?

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx [hunk 3]
```diff
@@ -272,11 +295,13 @@ export class TabBarToolbarRegistry implements FrontendApplicationContribution {
...
+            if (visible && (!item.when || this.contextKeyService.match(item.when, widget.node))) {
```
Comment (marechal-p) [inline packages/core/src/browser/shell/tab-bar-toolbar.tsx]:
```ts
            if (visible && item.when && this.contextKeyService.match(item.when, widget.node)) {
```

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx [hunk 4]
```diff
@@ -187,6 +187,24 @@ export interface TabBarToolbarItem {
 
 }
 
+/**
+ * Tab-bar toolbar item backed by a `React.ReactNode`.
+ * Unlike the `TabBarToolbarItem`, this item is not connected to the command service.
+ */
+export interface ReactTabBarToolbarItem {
+    readonly id: string;
+    readonly node: React.ReactNode;
```
Comment (akosyakov) [inline packages/core/src/browser/shell/tab-bar-toolbar.tsx]: How can one trigger rerendering?

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 2 above)
Comment (kittaakos) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: I think the current change is correct. You either have the `isVisible` defined, then it has to be evaluated to `true`, or you do not have this constraint; hence your item will always be visible.

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 3 above)
Comment (kittaakos) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: See https://github.com/theia-ide/theia/pull/4993#discussion_r278402449

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 1 above)
Comment (kittaakos) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: A [`Fragment` IS_A `ReactNode`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7c77a9c53a5d3e3c3c0a29a466b5acf810cac7a4/types/react/index.d.ts#L194).

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 4 above)
Comment (kittaakos) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: I do not think I'd need it, but I can add it to the API.

Any suggestions, by the way?

Comment (akosyakov) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: If you would like to render it differently depending on user input. As we do for usual items in the search widget.

>  Any suggestions, by the way?

There has to be an event like `onDidChange` and then it should not be `node` but a function which produces a new node, i.e. `render(): React.ReactNode`.

**Comment by kittaakos** [comment]
> How can one trigger rerendering?

I have updated the PR with this. Here is my dummy example:

```tsx
const emitter = new Emitter<void>();
const items = [1];
registry.registerItem({
    id: 'demo',
    render: () => <select
        name={`${items[items.length - 1]}`}
        value={`${items[items.length - 1]}`}
        title={`tooltip: ${items[items.length - 1]}`}>
        {items.map(option => <option value={option} key={option}>{option}</option>)}
    </select>,
    isVisible: widget => {
        if (widget instanceof EditorWidget) {
            return widget.editor.uri.toString().endsWith('.tsx');
        }
        return false;
    },
    onDidChange: emitter.event
});
window.setInterval(() => {
    items.push(items[items.length - 1] + 1);
    emitter.fire(undefined);
}, 3000);
```

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 2 above)
Comment (marechal-p) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: Ah ok, I made it default to false on my side, read that a bit quickly, you are right :)

File: packages/core/src/browser/shell/tab-bar-toolbar.tsx (hunk 1 above)
Comment (marechal-p) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: Hm, I was more wondering why we don't do the following:
```ts
    protected render(): React.ReactNode {
        return [...this.items.values()].map(item =>
            TabBarToolbarItem.is(item) ? this.renderItem(item) : item.render());
    }
```

Comment (kittaakos) [reply packages/core/src/browser/shell/tab-bar-toolbar.tsx]: I do not know; if you feel uncomfortable with the current code, please file an issue.

**Comment by marechal-p** [comment]
Finally managed to test, sorry it took long, I got myself into build troubles.

It works well, although I cannot change the <select> value when on Ubuntu/Firefox, but the element is here and the setInterval works.

**Pull request merged**